Tar decompression throws "permission denied" when archive is readonly

### Version

0.6.1

### Description

ouch throws permission error for read-only tar files. **(all the files/dirs in the tar are also readonly)**

### Current Behavior


```bash
❯ ouch d -y ./bash-uncompressed.tar
[INFO] Created temporary directory /home/USERNAME/Downloads/hi/./tmp-ouch-Zl5dmP to hold decompressed elements
[INFO] extracted (  0.00   B) "tmp-ouch-Zl5dmP"
[INFO] extracted (362.00   B) "tmp-ouch-Zl5dmP/b2bdb1880b6719555fa3e192ac77aa484e8cd256e1f27d678f5d43599a866651.json"
[INFO] extracted (  0.00   B) "tmp-ouch-Zl5dmP/bebcd3a5b3a2688d3bd968e2b54bc52208e0d66294ceafaebdeb5dde6afa779a"
[ERROR] failed to unpack `/home/USERNAME/Downloads/hi/./tmp-ouch-Zl5dmP/bebcd3a5b3a2688d3bd968e2b54bc52208e0d66294ceafaebdeb5dde6afa779a/VERSION`
 - Permission denied
```

### Expected Behavior

No errors.

### Additional Information

How to replicate:
* Create the archive:
    * ```nix-build -E 'with import <nixpkgs> {}; pkgs.dockerTools.buildImage { name = "bash-uncompressed"; tag = "latest"; compressor = "none"; copyToRoot = pkgs.bashInteractive; }'```
* ```cp -L result ./bash-uncompressed.tar```
* ```ls -ld ./bash-uncompressed.tar```  You will see `-r--r--r--`
* ```ouch d -y ./bash-uncompressed.tar```

--- a/src/archive/rar.rs
+++ b/src/archive/rar.rs
@@ -5,6 +5,7 @@ use std::path::Path;
 use unrar::Archive;
 
 use crate::{
+    commands::Unpacked,
     error::{Error, Result},
     list::FileInArchive,
     utils::{logger::info, Bytes},
@@ -17,7 +18,7 @@ pub fn unpack_archive(
     output_folder: &Path,
     password: Option<&[u8]>,
     quiet: bool,
-) -> crate::Result<usize> {
+) -> crate::Result<Unpacked> {
     let archive = match password {
         Some(password) => Archive::with_password(archive_path, password),
         None => Archive::new(archive_path),
@@ -43,7 +44,10 @@ pub fn unpack_archive(
         };
     }
 
-    Ok(unpacked)
+    Ok(Unpacked {
+        files_unpacked: unpacked,
+        read_only_directories: Vec::new(),
+    })
 }
 
 /// List contents of `archive_path`, returning a vector of archive entries

--- a/src/archive/sevenz.rs
+++ b/src/archive/sevenz.rs
@@ -12,6 +12,7 @@ use same_file::Handle;
 use sevenz_rust2::SevenZArchiveEntry;
 
 use crate::{
+    commands::Unpacked,
     error::{Error, FinalError, Result},
     list::FileInArchive,
     utils::{
@@ -98,7 +99,12 @@ where
     Ok(bytes)
 }
 
-pub fn decompress_sevenz<R>(reader: R, output_path: &Path, password: Option<&[u8]>, quiet: bool) -> crate::Result<usize>
+pub fn decompress_sevenz<R>(
+    reader: R,
+    output_path: &Path,
+    password: Option<&[u8]>,
+    quiet: bool,
+) -> crate::Result<Unpacked>
 where
     R: Read + Seek,
 {
@@ -167,7 +173,10 @@ where
         None => sevenz_rust2::decompress_with_extract_fn(reader, output_path, entry_extract_fn)?,
     }
 
-    Ok(count)
+    Ok(Unpacked {
+        files_unpacked: count,
+        read_only_directories: Vec::new(),
+    })
 }
 
 /// List contents of `archive_path`, returning a vector of archive entries

--- a/src/archive/tar.rs
+++ b/src/archive/tar.rs
@@ -3,6 +3,7 @@
 use std::{
     env,
     io::prelude::*,
+    ops::Not,
     path::{Path, PathBuf},
     sync::mpsc::{self, Receiver},
     thread,
@@ -12,27 +13,30 @@ use fs_err::{self as fs};
 use same_file::Handle;
 
 use crate::{
+    commands::Unpacked,
     error::FinalError,
     list::FileInArchive,
     utils::{
         self, create_symlink,
         logger::{info, warning},
-        Bytes, EscapedPathDisplay, FileVisibilityPolicy,
+        set_permission_mode, Bytes, EscapedPathDisplay, FileVisibilityPolicy,
     },
 };
 
 /// Unpacks the archive given by `archive` into the folder given by `into`.
 /// Assumes that output_folder is empty
-pub fn unpack_archive(reader: Box<dyn Read>, output_folder: &Path, quiet: bool) -> crate::Result<usize> {
+pub fn unpack_archive(reader: Box<dyn Read>, output_folder: &Path, quiet: bool) -> crate::Result<Unpacked> {
     let mut archive = tar::Archive::new(reader);
 
     let mut files_unpacked = 0;
+    let mut read_only_directories = Vec::new();
+
     for file in archive.entries()? {
         let mut file = file?;
 
         match file.header().entry_type() {
             tar::EntryType::Symlink => {
-                let relative_path = file.path()?.to_path_buf();
+                let relative_path = file.path()?;
                 let full_path = output_folder.join(&relative_path);
                 let target = file
                     .link_name()?
@@ -40,8 +44,26 @@ pub fn unpack_archive(reader: Box<dyn Read>, output_folder: &Path, quiet: bool) 
 
                 create_symlink(&target, &full_path)?;
             }
-            tar::EntryType::Regular | tar::EntryType::Directory => {
+            tar::EntryType::Regular => {
                 file.unpack_in(output_folder)?;
+            }
+            tar::EntryType::Directory => {
+                let original_mode = file.header().mode()?;
+                let is_writeable = (original_mode & 0o200) != 0;
+
+                file.unpack_in(output_folder)?;
+
+                if cfg!(unix) && is_writeable.not() {
+                    // We just unpacked a read-only directory
+                    // If any following entries are inside it (very likely), this would fail
+                    //
+                    // To get around that, we'll set this to writeable, then revert once finished
+                    let original_path = file.path()?.to_path_buf();
+                    let unpacked = output_folder.join(&original_path);
+                    set_permission_mode(&unpacked, original_mode | 0o200)?;
+
+                    read_only_directories.push((original_path, original_mode));
+                }
             }
             _ => continue,
         }
@@ -60,7 +82,10 @@ pub fn unpack_archive(reader: Box<dyn Read>, output_folder: &Path, quiet: bool) 
         files_unpacked += 1;
     }
 
-    Ok(files_unpacked)
+    Ok(Unpacked {
+        files_unpacked,
+        read_only_directories,
+    })
 }
 
 /// List contents of `archive`, returning a vector of archive entries

--- a/src/archive/zip.rs
+++ b/src/archive/zip.rs
@@ -17,6 +17,7 @@ use time::OffsetDateTime;
 use zip::{self, read::ZipFile, DateTime, ZipArchive};
 
 use crate::{
+    commands::Unpacked,
     error::FinalError,
     list::FileInArchive,
     utils::{
@@ -33,7 +34,7 @@ pub fn unpack_archive<R>(
     output_folder: &Path,
     password: Option<&[u8]>,
     quiet: bool,
-) -> crate::Result<usize>
+) -> crate::Result<Unpacked>
 where
     R: Read + Seek,
 {
@@ -120,7 +121,10 @@ where
         unpacked_files += 1;
     }
 
-    Ok(unpacked_files)
+    Ok(Unpacked {
+        files_unpacked: unpacked_files,
+        read_only_directories: Vec::new(),
+    })
 }
 
 /// List contents of `archive`, returning a vector of archive entries

--- a/src/commands/decompress.rs
+++ b/src/commands/decompress.rs
@@ -20,13 +20,21 @@ use crate::{
         io::lock_and_flush_output_stdio,
         is_path_stdin,
         logger::{info, info_accessible},
-        nice_directory_display, user_wants_to_continue,
+        nice_directory_display, set_permission_mode, user_wants_to_continue,
     },
     QuestionAction, QuestionPolicy, BUFFER_CAPACITY,
 };
 
 trait ReadSeek: Read + io::Seek {}
 impl<T: Read + io::Seek> ReadSeek for T {}
+
+pub type Mode = u32;
+pub type UnpackRes = crate::Result<Unpacked>;
+
+pub struct Unpacked {
+    pub files_unpacked: usize,
+    pub read_only_directories: Vec<(PathBuf, Mode)>,
+}
 
 pub struct DecompressOptions<'a> {
     pub input_file_path: &'a Path,
@@ -72,7 +80,10 @@ pub fn decompress_file(options: DecompressOptions) -> crate::Result<()> {
             Box::new(fs::File::open(options.input_file_path)?)
         };
         let zip_archive = zip::ZipArchive::new(reader)?;
-        let files_unpacked = if let ControlFlow::Continue(files) = execute_decompression(
+        let Unpacked {
+            files_unpacked,
+            read_only_directories: _,
+        } = if let ControlFlow::Continue(files) = execute_decompression(
             |output_dir| crate::archive::zip::unpack_archive(zip_archive, output_dir, options.password, options.quiet),
             options.output_dir,
             &options.output_file_path,
@@ -145,7 +156,7 @@ pub fn decompress_file(options: DecompressOptions) -> crate::Result<()> {
         reader = chain_reader_decoder(format, reader)?;
     }
 
-    let files_unpacked = match first_extension {
+    let Unpacked { files_unpacked, .. } = match first_extension {
         Gzip | Bzip | Bzip3 | Lz4 | Lzma | Xz | Lzip | Snappy | Zstd | Brotli => {
             reader = chain_reader_decoder(&first_extension, reader)?;
 
@@ -160,7 +171,10 @@ pub fn decompress_file(options: DecompressOptions) -> crate::Result<()> {
 
             io::copy(&mut reader, &mut writer)?;
 
-            1
+            Unpacked {
+                files_unpacked: 1,
+                read_only_directories: Vec::new(),
+            }
         }
         Tar => {
             if let ControlFlow::Continue(files) = execute_decompression(
@@ -213,8 +227,7 @@ pub fn decompress_file(options: DecompressOptions) -> crate::Result<()> {
         }
         #[cfg(feature = "unrar")]
         Rar => {
-            type UnpackResult = crate::Result<usize>;
-            let unpack_fn: Box<dyn FnOnce(&Path) -> UnpackResult> = if options.formats.len() > 1 || input_is_stdin {
+            let unpack_fn: Box<dyn FnOnce(&Path) -> UnpackRes> = if options.formats.len() > 1 || input_is_stdin {
                 let mut temp_file = tempfile::NamedTempFile::new()?;
                 io::copy(&mut reader, &mut temp_file)?;
                 Box::new(move |output_dir| {
@@ -311,13 +324,13 @@ pub fn decompress_file(options: DecompressOptions) -> crate::Result<()> {
 }
 
 fn execute_decompression(
-    unpack_fn: impl FnOnce(&Path) -> crate::Result<usize>,
+    unpack_fn: impl FnOnce(&Path) -> crate::Result<Unpacked>,
     output_dir: &Path,
     output_file_path: &Path,
     question_policy: QuestionPolicy,
     is_output_dir_provided: bool,
     is_smart_unpack: bool,
-) -> crate::Result<ControlFlow<(), usize>> {
+) -> crate::Result<ControlFlow<(), Unpacked>> {
     if is_smart_unpack {
         return smart_unpack(unpack_fn, output_dir, output_file_path, question_policy);
     }
@@ -336,10 +349,10 @@ fn execute_decompression(
 /// - If `output_dir` does not exist OR is a empty directory, it will unpack there
 /// - If `output_dir` exist OR is a directory not empty, the user will be asked what to do
 fn unpack(
-    unpack_fn: impl FnOnce(&Path) -> crate::Result<usize>,
+    unpack_fn: impl FnOnce(&Path) -> crate::Result<Unpacked>,
     output_dir: &Path,
     question_policy: QuestionPolicy,
-) -> crate::Result<ControlFlow<(), usize>> {
+) -> crate::Result<ControlFlow<(), Unpacked>> {
     let is_valid_output_dir = !output_dir.exists() || (output_dir.is_dir() && output_dir.read_dir()?.next().is_none());
 
     let output_dir_cleaned = if is_valid_output_dir {
@@ -367,11 +380,11 @@ fn unpack(
 ///
 /// Note: This functions assumes that `output_dir` exists
 fn smart_unpack(
-    unpack_fn: impl FnOnce(&Path) -> crate::Result<usize>,
+    unpack_fn: impl FnOnce(&Path) -> crate::Result<Unpacked>,
     output_dir: &Path,
     output_file_path: &Path,
     question_policy: QuestionPolicy,
-) -> crate::Result<ControlFlow<(), usize>> {
+) -> crate::Result<ControlFlow<(), Unpacked>> {
     assert!(output_dir.exists());
     let temp_dir = tempfile::Builder::new().prefix("tmp-ouch-").tempdir_in(output_dir)?;
     let temp_dir_path = temp_dir.path();
@@ -381,7 +394,10 @@ fn smart_unpack(
         nice_directory_display(temp_dir_path)
     ));
 
-    let files = unpack_fn(temp_dir_path)?;
+    let Unpacked {
+        files_unpacked,
+        read_only_directories,
+    } = unpack_fn(temp_dir_path)?;
 
     let root_contains_only_one_element = fs::read_dir(temp_dir_path)?.take(2).count() == 1;
 
@@ -410,11 +426,26 @@ fn smart_unpack(
     if fs::rename(&previous_path, &new_path).is_err() {
         utils::rename_recursively(&previous_path, &new_path)?;
     };
+
+    if cfg!(unix) {
+        for (path, mode) in &read_only_directories {
+            let components = path.components();
+            let mut path = new_path.clone();
+            for component in components {
+                path.push(component);
+            }
+            set_permission_mode(&path, *mode)?;
+        }
+    }
+
     info_accessible(format!(
         "Successfully moved \"{}\" to \"{}\"",
         nice_directory_display(&previous_path),
         nice_directory_display(&new_path),
     ));
 
-    Ok(ControlFlow::Continue(files))
+    Ok(ControlFlow::Continue(Unpacked {
+        files_unpacked,
+        read_only_directories,
+    }))
 }

--- a/src/commands/mod.rs
+++ b/src/commands/mod.rs
@@ -8,6 +8,7 @@ use std::{ops::ControlFlow, path::PathBuf};
 
 use bstr::ByteSlice;
 use decompress::DecompressOptions;
+pub use decompress::Unpacked;
 use rayon::prelude::{IndexedParallelIterator, IntoParallelRefIterator, ParallelIterator};
 use utils::colors;
 

--- a/src/utils/fs.rs
+++ b/src/utils/fs.rs
@@ -243,3 +243,19 @@ pub fn create_symlink(target: &Path, full_path: &Path) -> crate::Result<()> {
 
     Ok(())
 }
+
+#[cfg(unix)]
+#[inline]
+pub fn set_permission_mode(path: &Path, mode: u32) -> crate::Result<()> {
+    use std::{fs::Permissions, os::unix::fs::PermissionsExt};
+
+    fs::set_permissions(path, Permissions::from_mode(mode))?;
+
+    Ok(())
+}
+
+#[cfg(windows)]
+#[inline]
+pub fn set_permission_mode(path: &Path, mode: u32) -> crate::Result<()> {
+    Ok(())
+}

--- a/src/utils/mod.rs
+++ b/src/utils/mod.rs
@@ -19,7 +19,8 @@ pub use self::{
     },
     fs::{
         cd_into_same_dir_as, create_dir_if_non_existent, create_symlink, is_path_stdin, remove_file_or_dir,
-        rename_for_available_filename, rename_recursively, resolve_path_conflict, try_infer_extension,
+        rename_for_available_filename, rename_recursively, resolve_path_conflict, set_permission_mode,
+        try_infer_extension,
     },
     question::{
         ask_to_create_file, user_wants_to_continue, user_wants_to_overwrite, FileConflitOperation, QuestionAction,

